packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,542 @@
+    final Target target = _getNamedTarget(name);
+    environment.buildDir.createSync(recursive: true);
+
+    // Load file cache from previous builds.
+    final FileCache fileCache = FileCache(environment)
+      ..initialize();
+
+    // Perform sanity checks on build.
+    checkCycles(target);
+
+    // TODO(jonahwilliams): create a separate configuration for the settings and
+    // constants which effect build running.
+    final _BuildInstance buildInstance = _BuildInstance(environment, fileCache, buildSystemConfig);
+    try {
+      await buildInstance.invokeTarget(target);
+    } finally {
+      // Always persist the file cache to disk.
+      fileCache.persist();
+    }
+  }
+
+  /// Describe the target `name` and all of its dependencies.
+  List<Map<String, Object>> describe(
+    String name,
+    Environment environment,
+  ) {
+    final Target target = _getNamedTarget(name);
+    checkCycles(target);
+    // Cheat a bit and re-use the same map.
+    Map<String, Map<String, Object>> fold(Map<String, Map<String, Object>> accumulation, Target current) {
+      accumulation[current.name] = current.toJson(environment);
+      return accumulation;
+    }
+
+    final Map<String, Map<String, Object>> result =
+        <String, Map<String, Object>>{};
+    final Map<String, Map<String, Object>> targets = target.fold(result, fold);
+    return targets.values.toList();
+  }
+
+  // Returns the corresponding target or throws.
+  Target _getNamedTarget(String name) {
+    final Target target = targets
+        .firstWhere((Target target) => target.name == name, orElse: () => null);
+    if (target == null) {
+      throw Exception('No registered target named $name.');
+    }
+    return target;
+  }
+}
+
+/// An active instance of a build.
+class _BuildInstance {
+  _BuildInstance(this.environment, this.fileCache, this.buildSystemConfig)
+    : resourcePool = Pool(buildSystemConfig.resourcePoolSize ?? platform?.numberOfProcessors ?? 1);
+
+  final BuildSystemConfig buildSystemConfig;
+  final Pool resourcePool;
+  final Map<String, AsyncMemoizer<void>> pending = <String, AsyncMemoizer<void>>{};
+  final Environment environment;
+  final FileCache fileCache;
+
+  Future<void> invokeTarget(Target target) async {
+    await Future.wait(target.dependencies.map(invokeTarget));
+    final AsyncMemoizer<void> memoizer = pending[target.name] ??= AsyncMemoizer<void>();
+    await memoizer.runOnce(() => _invokeInternal(target));
+  }
+
+  Future<void> _invokeInternal(Target target) async {
+    final PoolResource resource = await resourcePool.request();
+    try {
+      final List<SourceFile> inputs = target.resolveInputs(environment);
+      final Map<String, ChangeType> updates = await target.computeChanges(inputs, environment, fileCache);
+      if (updates.isEmpty) {
+        printTrace('Skipping target: ${target.name}');
+      } else {
+        printTrace('${target.name}: Starting');
+        await target.invocation(updates, environment);
```

Comment:
Note to test the case that this call throws an exception.

Reply:
Done